File: C++/1261. Find Elements in a Contaminated Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class FindElements {
public:
    set<int> s;

    void dfs(TreeNode* node) {
        if(!node) return;

        if(node->left){
            node->left->val = node->val * 2 + 1;
            s.insert(node->left->val);
            dfs(node->left);
        }
        if(node->right){
            node->right->val = node->val * 2 + 2;
            s.insert(node->right->val);
            dfs(node->right);
        }
    }

    FindElements(TreeNode* root) {
        root->val = 0;
        s.insert(0);
        dfs(root);
    }
    
    bool find(int target) {
        return s.count(target);
    }
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */

File: Python/1261. Find Elements in a Contaminated Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class FindElements:

    def __init__(self, root: Optional[TreeNode]):
        root.val = 0
        self.s = set()
        self.s.add(0)

        def dfs(node):
            if not node: return

            if node.left:
                node.left.val = node.val * 2 + 1
                self.s.add(node.left.val)
                dfs(node.left)
            if node.right:
                node.right.val = node.val * 2 + 2
                self.s.add(node.right.val)
                dfs(node.right)

        dfs(root)

    def find(self, target: int) -> bool:
        return target in self.s


# Your FindElements object will be instantiated and called as such:
# obj = FindElements(root)
# param_1 = obj.find(target)
